content/blog/2022-06-28-serving-models-with-mlem.md

```diff
@@ -0,0 +1,277 @@
+  ├── 75b1b3c [exp-2ba75]    0.67913   0.38207   20210428     350           75
+  └── 40b0075 [exp-e0dc1]    0.76681   0.38867   20210428     300           75
+ ───────────────────────────────────────────────────────────────────────────────────────────
+```
+
+Of course you would run many more experiments in order to determine the best
+model for production, but in this example, the metrics for `exp-5ebf6` look
+pretty good. So we'll take this model and get it ready to serve to production.
+
+## Save the model
+
+First, we need to install the `MLEM` package with:
+
+```dvc
+$ pip install 'mlem[fastapi]'
+```
+
+Then, inside of the `train.py` script, we need to add the MLEM import to save
+the models as we experiment.
+
+```python
+# train.py
+
+import os
+import pickle5 as pickle
+import sys
+import yaml
+
+from mlem.api import save
+import numpy as np
+from sklearn.ensemble import RandomForestClassifier
+...
+```
+
+Then you can add the `save` function to the end of the training script.
+
+```python
+# train.py
+...
+clf = RandomForestClassifier(
+    n_estimators=n_est, min_samples_split=min_split, n_jobs=2, random_state=seed
+)
+
+clf.fit(x, labels)
+
+save(
+    clf,
+    "clf",
+    sample_data=x,
+    description="Random Forest Classifier",
+)
+```
+
+_Note:_ If you want to see what is happening in the DVC pipeline, check out the
+`dvc.yaml` file in the project.
+
+There will be two new files in your repo: `clf` and `clf.mlem`. You'll be able
+to load the model file into the evaluation script and make sure it gives you the
+performance you need.
+
+## Load the model for evaluation
+
+We're going to load the MLEM model into our `src/evaluate.py` script. To do
+that, you'll need to add the following import.
+
+```python
+# evaluate.py
+
+import json
+import math
+import os
+import sys
+
+import pickle5 as pickle
+import sklearn.metrics as metrics
+import numpy as np
+from mlem.api import apply
+...
+```
+
+Now we can use the [`apply` function](https://mlem.ai/doc/api-reference/apply)
+to make predictions with the model.
+
+```python
+# evaluate.py
+...
+x = matrix.iloc[:,1:11].values
+
+cleaned_x = np.where(np.isnan(x), 0, x)
+labels_pred = apply(model_file, cleaned_x, method="predict")
+
+predictions_by_class = apply(model_file, cleaned_x, method="predict_proba")
+predictions = predictions_by_class[:, 1]
+...
+```
+
+The `predict` and `predict_proba` are methods available from the model and they
+are used to get new predicted values and their probabilities for evaluation.
+This, alond with everything else in the script, is how we get those metrics are
+shown in the experiments table. Once you've validated the model and decided that
+it performs well, it's ready to go to production.
+
+## Serve the model to production
+
+There are a couple of ways you can do this with MLEM:
+
+- Create a Python package
+- Serve the model with [FastAPI](https://fastapi.tiangolo.com/)
```

Comment:
1. Takes a while to get here. Maybe we should skip the whole training?
2. The title only mentioned the web API but we're also introducing the Python package here. OK to keep but after `mlem serve` I think, and include in title or at least post descriptions, I think.

Reply:
1. Yeah, I made the training/evaluate steps an info box that can be skipped.
2. I moved the Python package to make it the second way to serve a model.